Csiro/s3 driver

petewa opened the pull request:
### Reason for this pull request

Improvements to DataCube:
   - S3 storage backend
      - Windows is not supported.
   - driver manager to dynamically load/switch storage drivers (NetCDF, S3, S3-test).
   - Ingest now supports creation of nD Storage Units in the available storage drivers.
   	 e.g.
   	 	NetCDF:  datacube -v ingest -c ls5_nbar_albers.yaml --executor multiproc 8
		S3:      datacube --driver s3 -v ingest -c  ls5_nbar_albers_s3.yaml --executor multiproc 8
		S3-test: datacube --driver s3-test -v ingest -c ls5_nbar_albers_s3_test.yaml --executor multiproc 8
  - load and load_data has optional multi-threaded support with use_threads flag.

Improved testing:
 - tests are run for each driver, where possible.
 - tests corner values
 - tests md5 hash equality on load_data
 - tests multiple time slices
 - reduction in data usage
 - reduction in total number of concurrent db connections.

### Proposed changes
 - cli driver parameter to select driver, if None, it defaults to NetCDF.
 - support for generating n-dimension storage units on ingest.
   - example ingest yaml: docs/config_samples/ingester/ls5_nbar_nbar_s3.yaml
 - supported drivers:
   - NetCDF: based on existing driver.
   - S3: S3 backend for storage.
   - S3-test: Same as S3 but emulated on disk.
 - datacube.api.load_data - use_threads parameter to enable threaded _fuse_measurement with results stored in a shared memory array.
 - datacube.scripts.ingest - uses slightly modified GridWorkFlow to generate 3D tasks for 3D Storage Unit creation.
 - optional creation of s3 tables via "datacube -v system init -s3"

### Todo:
 - More tests.

 - [ ] Closes #xxxx
 - [ ] Tests added / passed
 - [ ] Fully documented, including `docs/about/whats_new.rst` for all changes

<!--
See https://github.com/blog/2111-issue-and-pull-request-templates for more
information on pull request templates.

-->

Commit:
Initial Push for the CSIRO S3 Driver

Created directory structure for Index & Storage
Index
Storage
 S3AIO - S3 Array IO
 S3IO - S3 IO

Commit:
Added boto3, SharedArray to conda environments for travis.

rtaib committed:
Temporary changes to integration tests to be able to run simple ingestions.

rtaib committed:
Logging and driver switching in ingest. For testing

rtaib committed:
Corrected formatting to please pylint.

rtaib committed:
Reverted to csiro s3 branch until I know how to test integration on my local vm :(

petewa committed:
S3IO
 - option to specify directory for s3 storage units

S3AIO
 - option to specify directory for s3 storage units

rtaib committed:
Started to edit the tests to add an s3 switch, and started to introduce it into the ingester.

rtaib committed:
Added support for multiple observations in full ingestion test. This will help test t,x,y ingestion with the s3 driver. For now, a single dataset is used, though.

petewa committed:
Initial commit for discussion.

datacube/api/grid_workflow.py
 - Tile.split now supports time chunking.

datacube/scripts/ingest.py
 - now supports 3D ingest via ingest config yaml
   e.g.

    storage:
      driver: NetCDF CF

      crs: EPSG:3577
      tile_size:
              x: 100000.0
              y: 100000.0
              time: 20
      resolution:
              x: 25
              y: -25
      chunking:
              x: 200
              y: 200
              time: 1
      dimension_order: ['time', 'y', 'x']

TODO: testing

petewa committed:
Fixed travis error:
  - Comited wrong ingest.py file.

petewa committed:
Make sharding arrays to S3 simpler.
  e.g.
    import datacube.drivers.s3.storage.s3aio as s3aio
    import numpy as np
    s = s3aio.S3AIO()
    x = np.arange(4*4*4,dtype=np.uint8).reshape((4,4,4))
    key_map = s.put_array_in_s3(x, (2,2,2), "base_name", 'arrayio')

petewa committed:
datacube/drivers/s3/storage/s3aio/__init__.py
datacube/drivers/s3/storage/s3aio/s3lio.py
  - Added S3 Labelled Array IO class
    - labelled array reads/writes that spans multiple s3 objects.
    - stubs for get_data

datacube/drivers/s3/storage/s3aio/s3aio.py
  - moved functionality that spans multiple s3 objects to s3lio
  - function to return s3io
  - comments

datacube/drivers/s3/storage/s3aio/s3io.py
  - comments

rtaib committed:
Remove duplicates within cells in ingester. The comparison between cells to be ingested and cells already indexed is made on the tuple (extent, timestamp). Using the dataset ID did not seem to be an option as it is assigned randomly to each new tile to be ingested.

petewa committed:
datacube/drivers/s3/storage/s3aio/s3lio.py
  - Support for bounded array chunking.

petewa committed:
datacube/drivers/s3/storage/s3aio/s3lio.py
  - prefix keys with first 6 bytes of md5 hash.

rtaib committed:
Created a simple plugin manager for storage drivers. For the moment, it is mostly a placeholder, but it will soon receive the s3 driver.

rtaib committed:
Integrated the driver loader in the current storage module, and now calling it from ingest. A skeleton s3 implementation was created, but proper unpacking of the dataset and parameters needs discussing.

rtaib committed:
Added missing __init__.py which caused an import error in python 2.7. Temporarily disable pylint on some issues in the s3aio classes.

rtaib committed:
Temporarily disabled additional pylint issues, reported in python 2.7

rtaib committed:
More python 2.7 pylint fixes...

petewa committed:
Fix pylint errors.

petewa committed:
datacube/scripts/ingest.py
  - Added driver to output_type.definition['storage']

petewa committed:
Changed order of parameters

petewa committed:
integration_tests/conftest.py
  - Fixed alter_dataset_type_for_testing call in ls5_nbar_ingest_config

Commit:
iInitial work on database schema for S3 objects

Commit:
Space...

Commit:
Spaces again...

petewa committed:
datacube/drivers/s3/storage/s3aio/s3io.py:
  - Added function to delete keys from a bucket.

rtaib committed:
S3 driver now writing data to s3. Added an s3-test driver to allow travis to write to filesystem instead of s3, hence avoiding credentials protection issues. Updated tests accordingly.

rtaib committed:
Merge branch 'csiro/s3-driver' of https://github.com/opendatacube/datacube-core into csiro/s3-driver

rtaib committed:
De-activating s3 test until travis disc space gets discussed.

rtaib committed:
Fixed memory issue in python2.7 due to incompatibilities between `map` and `map_async` using different types of iterators in each python version. Also dramatically reduced the size of the test data (empty geotiffs) to save disk space. This may need reviewing to match the yaml data somehow.

rtaib committed:
Made driver loader a manager. It is now a singleton that can be used to call methods into the drivers.

petewa committed:
datacube/drivers/s3/storage/s3aio/s3lio.py
 - Fixed commit 842a764
   - not entering process worker in python 2.7
 - Moved to pathos multiprocessing
 - Python 2.7 compatibility

.travis/environment_py27.yaml
.travis/environment_py35.yaml
 - Added pathos multiprocessing

petewa committed:
datacube/drivers/s3/storage/s3aio/s3aio.py
datacube/drivers/s3/storage/s3aio/s3io.py
datacube/drivers/s3/storage/s3aio/s3lio.py
  - move to pathos multiprocessing
  - python 2.7 compability

rtaib committed:
Added a global `driver` option to the command-line, which initialises the index through the driver manager. It all defaults to the NetCDF driver and Postgres index.

rtaib committed:
Merge branch 'csiro/s3-driver' of https://github.com/opendatacube/datacube-core into csiro/s3-driver

Commit:
Merge branch 'csiro/s3-driver' into csiro/s3-driver-database-new

Commit:
Add initial insert queries for S3 storage

petewa committed:
setup.py
  = Added pathos

Commit:
Generate UUIDs in Python

Commit:
Add S3 tables to tables init

rtaib committed:
DriverManager now sets a current driver when it gets (re-)instantiated. The current driver is initialised along with its index and is used as default driver in any subsequent DriverManager call. For now, this happens in the click interface, which requires a --driver flag. This align with the idea of removing driver information from the yaml config files. It will need to be added to the datacube initialisation to allow end-users to specify the driver/index corresponding to their datasets.

rtaib committed:
Merge branch 'csiro/s3-driver' of https://github.com/opendatacube/datacube-core into csiro/s3-driver

rtaib committed:
Merge remote-tracking branch 'origin/csiro/s3-driver-database-new' into csiro/s3-driver

petewa committed:
datacube/drivers/s3/storage/s3aio/s3lio.py
  - added chunk_id to return in put_array_to_s3()

rtaib committed:
Merge branch 'csiro/s3-driver' of https://github.com/opendatacube/datacube-core into csiro/s3-driver

rtaib committed:
Added `base_name` column to S3_DATASET table.

rtaib committed:
Removed unwanted whitespaces to please pylint.

rtaib committed:
S3 indexer implemented.

rtaib committed:
Integrated drivers documentation to main documentation by adding RST scripts in docsand updating config where needed.

rtaib committed:
Removed a trailing whitespace.

rtaib committed:
Started to implement the s3 data retrieval. DatasetSource is now retrieved through the driver manager. The s3 version is not functional yet.

rtaib committed:
Started implementation of data access using s3. The driver manager was rewritten in depth to load all drivers so the appropriate driver gets used to retrieve each dataset, based on its location specified in the DB (or config file). This change has ramifications in several classes, including tests. The manager is still a singleton for now, but holds a reference to a default driver, which is only used if data comes without driver specified for its ingestion. The mechanisms to pipe requests from core to actual storage queries are not implemented yet, in particular the DB retrieval of s3 parameters.

petewa committed:
Merge branch 'develop' into csiro/s3-driver

petewa committed:
datacube/drivers/s3/index.py
  - modified _make to reflect changes from develop.

petewa committed:
datacube/drivers/s3/index.py
  - commented problem line writing to dataset.local_uri
    - Todo - adapt for dataset.uris

rtaib committed:
Extracted s3 metadata from DB and added to the dataset as `index_specific` information. This code will need to be restructured, with parts moving into the driver.

petewa committed:
datacube/drivers/s3/storage/s3aio/s3lio.py
  - added get_data_unlabelled function to retrieve array data sharded across S3 objects.

datacube/drivers/s3/storage/s3aio/s3aio.py
  - fix num_streams divisor.

petewa committed:
datacube/drivers/s3/storage/s3aio/s3lio.py
  - Fix travis

rtaib committed:
Moved the s3 index functionality completely inside the driver. Edited the integration conftest to make drivers available as a fixture. The driver manager now keeps the same default driver even if the index gets changed. Updated all tests where required to use flexible uri schemes instead of hardcoding to file.

petewa committed:
datacube/drivers/s3/storage/s3aio/s3lio.py
  - modified to use origin/shape when computing local slices

rtaib committed:
Add dataset script now uses the driver manager to determine the appropriate driver to index a dataset, based on its paths.

rtaib committed:
Fixed invalid uri scheme for s3 driver (by Peter).

rtaib committed:
Geotiff used in integration tests now contain predictable values (by Peter).

rtaib committed:
Implemented s3 datasource, that retrieves data from storage. Moved `bucket` metadata from chunk table to s3 dataset table

rtaib committed:
S3 datasource now reuses the driver storage. Cleaned up and documented the S3 datasource module.

rtaib committed:
Modified conftest to create multiple time slices based on the existing sample data. Modified test_full_ingestion accordingly. Started to test for specific content in test_full_ingestion (although only checking for 0 for now)

rtaib committed:
In conftest, now creating a unique geotiff per band and symlink it from all tests in session, to save generation time and disk space (down from 15GB to 503MB).

rtaib committed:
Modified conftest to allow flexible size geotiffs to be created. This makes for faster tests and specific data checks (for now, specific values in the corners of the geotiff). The corresponding test was implemented in test_full_ingestion but is not working yet because of required changes in s3/datasouce. A minor temporary fix was added, but more work is underway.

petewa committed:
datacube/drivers/s3/datasource.py
 - minor fixes

petewa committed:
Revert back to netcdf 1.2.7.

rtaib committed:
Moved epsilon constant to a utils module, shared by all drivers. It is used as error margin where dimension comparisons are needed, e.g. to determine a regular or irregular time index.

rtaib committed:
Test full ingestion now checks specific values in the corners of a geotiff. The sclaing factor is reduced during the process, from geotiff creation in conftest to testing in full ingestion, but the crs remains fixed to avoid any warping.

rtaib committed:
Improved tests to use multiple time slices and flexible size.

petewa committed:
datacube/api/core.py
 - Added support for multi-thread reads/call to _fuse_measurement
   - defaults to off unless use_threads=True

datacube/drivers/s3/driver.py
 - set S3LIO to use s3

datacube/drivers/s3/storage/s3aio/s3lio.py
datacube/drivers/s3/storage/s3aio/s3aio.py
datacube/drivers/s3/storage/s3aio/s3io.py
 - Added support for compression.
 - minor refactor

.travis/environment_py27.yaml
.travis/environment_py35.yaml
setup.py
 - added zstandard

petewa committed:
datacube/drivers/s3/datasource.py
  - Use multi-process version of get_data_unlabeled

petewa committed:
datacube/drivers/s3/index.py
  - commenting unused code for now. Need to revisit.

rtaib committed:
Added a generic index that looks up basic dataset info in the DB, before handing over to the driver-specific drivers to add any specific information required.

rtaib committed:
Merge branch 'csiro/s3-driver' of https://github.com/opendatacube/datacube-core into csiro/s3-driver

rtaib committed:
Made driver manager a normal object instance instead of a singleton, allowing to manage index connections separately between different objects using them, e.g. tests. So far, it seems that connections were left open until they times out, peaking at about 35 in the develop branch, and 70 in the s3 driver branch, since it doubles all tests. Using separate driver instances and closing it each time reduced the total number of concurrent connections to 9 max in the complete test suite. The downside is a possible speed reduction in tests, but that should not apply to normal usage of the system since a single driver could be used throughout.

rtaib committed:
Updated docstrings to reflect the introduction of driver manager as an object.

rtaib committed:
Removed a couple of spurious logging blocks, and a test switch. Updated driver manager doc to reflect it is now an object not a singleton. Renamed "default" driver into "current" driver throughout, to better indicate this may be changed during usage.

petewa committed:
Merge branch 'develop' into csiro/s3-driver

petewa committed:
docs/config_samples/ingester/ls5_nbar_albers_s3.yaml
 - Sample s3 ingest config to ingest to s3.

docs/config_samples/ingester/ls5_nbar_albers_s3test.yaml
 - Sample s3-test ingest config to ingest to emulated s3 on disk. *for testing only*

petewa committed:
docs/conf.py
 - Added missing comma.

petewa committed:
Added Documentation.

petewa committed:
Added to whats_new.

rtaib committed:
Minor changes in view of pylint 1.7 compatibility. However, pylint was fixed to 1.6.5 for now.

petewa committed:
datacube/drivers/s3/storage/s3aio/s3aio.py
datacube/drivers/s3/storage/s3aio/s3io.py
datacube/drivers/s3/storage/s3aio/s3lio.py
datacube/drivers/s3_test/driver.py
  - add compression flag

datacube/drivers/manager.py
  - internal pickling

tests/drivers/test_s3_storage.py
tests/analytics_execution_engine/test_analyics_execution_engine.py
  - Added tests

tests/api/test_api.py
tests/api/test_grid_workflow.py
  - Make mock picklable

datacube/scripts/ingest.py
  - added driver_manager.close() when cloned.

petewa committed:
Merge branch 'csiro/s3-driver' of github.com:opendatacube/datacube-core into csiro/s3-driver

petewa committed:
datacube/drivers/s3/storage/s3aio/s3lio.py
  - fix py2 issue

tests/drivers/test_s3_storage.py
  - fixed and added test

omad committed:
Formatting fixes

- Follow PEP8 for spacing between classes/functions
- Follow PEP257 for docstring (double quotes not single)

omad committed:
More Formatting fixes

- Follow PEP8 for spacing between classes/functions
- Follow PEP257 for docstring (double quotes not single)

petewa committed:
Remove space.

Kirill888 commented on datacube/drivers/index.py:
```diff
@@ -0,0 +1,87 @@
+          index on initialisation. Caution: In the current
+          implementation all parameters get passed to all potential
+          indexes.
+        :param args: Optional keyword arguments to be passed to the
```
should be `:params kargs:` (missing `k`)

omad committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	.travis/environment_py27.yaml
	datacube/api/_api.py
	datacube/execution/execution_engine.py
	datacube/scripts/ingest.py
	datacube/scripts/system.py
	datacube/storage/storage.py
	datacube/ui/click.py
	docs/about/whats_new.rst
	integration_tests/conftest.py
	integration_tests/index/test_config_docs.py
	setup.py

omad committed:
Merge remote-tracking branch 'origin/csiro/s3-driver' into csiro/s3-driver

omad committed:
Merge branch 'develop' into csiro/s3-driver

omad committed:
Bump pylint version

omad committed:
Fix travis pylint req

omad committed:
Fix merge breakages

omad committed:
Remove checks absent in pylint 1.7.2

omad commented on datacube/drivers/driver.py:
```diff
@@ -0,0 +1,116 @@
+        """
+        path = str(path)
+        body = path.split(':', 1)[1] if ':' in path else path
+        return '%s:%s' % (self.uri_scheme, body)
```
This is overly simplistic URI generation, in particular, local file URLs should start like `file:///myrootfile.txt` not `file:/myrootfile.txt'. There is code elsewhere that handles this, we'll just need to plumb it through.

omad commented on datacube/drivers/manager.py:
```diff
@@ -0,0 +1,316 @@
+import weakref
+from pathlib import Path
+from collections import Iterable
+from dill import loads, dumps
```
It would be nice to test whether `cloudpickle` will work instead of `dill`. I imagine it does, and we already depend on it via `xarray/dask`.

omad commented on datacube/drivers/manager.py:
```diff
@@ -0,0 +1,316 @@
+
+# Dynamic loading from filename varies across python versions
+# Based on http://stackoverflow.com/a/67692
+if sys.version_info >= (3, 5):  # python 3.5+
```
I love that there's support for multiple python versions! We normally try and keep all of this code in a central `datacube/compat.py` file and import required functions from there.

omad commented on datacube/drivers/manager.py:
```diff
@@ -0,0 +1,316 @@
+    """
+
+    _DEFAULT_DRIVER = 'NetCDF CF'
+    '''Default 'current' driver, assuming its code is present.
```
Instead of 'docstrings' for class attributes, the convention is to use
```python
    #: Default 'current' driver, assuming its code is present.
    _DEFAULT_DRIVER = 'NetCDF CF'
```

omad commented on datacube/index/postgres/tables/_schema.py:
```diff
@@ -111,3 +112,92 @@
+
+
+# --- S3-driver specific Tables ---
+S3_DATASET = Table(
```
I had expected the S3 support to be optional, ie. only create new database tables if it's enabled.

It's not a huge problem having extra tables always being created, but optional would have been nicer.

omad commented on datacube/drivers/manager.py:
```diff
@@ -0,0 +1,316 @@
+                    driver.index.close()
+        self.__drivers = {}
+        for init_path in Path(__file__).parent.glob('*/__init__.py'):
+            init = load_module(str(init_path.parent.stem), str(init_path))
```
Around here could deal with `ImportErrors` with drivers, when they depend on packages which aren't installed. Drivers could also check that any other pre-requirements are dealt with, ie. if necessary Database tables were available for use, and fail silently with a warning if not.

omad commented on datacube/index/postgres/_api.py:
```diff
@@ -868,3 +873,124 @@ def grant_role(self, role, users):
+
+    # S3 specific functions
+    # See .tables for description of each column
+    def put_s3_mapping(self, dataset_ref, band, s3_dataset_id):
```
Still not convinced by s3 specifics polluting this already crowded class. It would be great to have these separate. If we get more drivers the problem will only become worse.

omad commented on datacube/api/core.py:
```diff
@@ -1,21 +1,26 @@
 import pandas
 import numpy
 import xarray
+import SharedArray as sa
```
I'd really like `SharedArray` to be an optional dependency. It's currently only available as a source package requiring compilation, which makes installing it much more difficult on Windows/Mac systems without a compiler.

It's also slowed down the review of all this work, since none of the existing environments available to me could run this branch.

omad reviewed (changes requested):
Looks great guys, Love all the documentation and testing, looks like high quality code.

There are however a couple of issues, a couple of which I consider blockers:
- [ ] Mandatory dependency on `SharedArray`
- [ ] Slightly erroneous URI handling, which is the cause of the current test failures.

And a few minor things which shouldn't be hard to correct:
- [ ] Use `cloudpickle` instead of `dill`, since we already depend on it and afaik they do the same job.
- [ ] Move multi-version python compatible code into the `compat` module.

petewa committed:
datacube/drivers/driver.py
 - fixed uri for tests, this function is only used in tests

datacube/drivers/index.py
 - added missing k

integration_tests/index/test_index_data.py
 - fixed tests to reflect other uri schemes contained in drivers.

petewa committed:
datacube/compat.py
 - moved python version specific load_module import here from driver manager

datacube/drivers/manager.py
 - removed python version specific load_module import
 - fixed docstrings

datacube/drivers/s3/index.py
 - moved s3 specific functions here from datacube.index.postgres._api

datacube/index/postgres/_api.py
 - created execute function to pass in sqlalchemy commands to be executed.
 - removed  moved s3 specific functions

petewa committed:
datacube/api/core.py
 - optional import of SharedArray
 - added docstrings for use_threads in load and load_data

datacube/drivers/driver.py
 - driver requirement check.

datacube/drivers/manager.py
 - check driver requirement and silent fail with warning.
 - check driver for failed load and silent fail with warning.
 - change from dill to cloud pickle (dill can handle more e.g. dill can pickle the datacube instance)

datacube/drivers/s3/driver.py
 - driver requirement check.

datacube/index/_api.py
datacube/index/postgres/_connections.py
datacube/index/postgres/tables/_core.py
datacube/index/postgres/tables/_schema.py
datacube/scripts/system.py
integration_tests/conftest.py
 - optional creation of s3 tables

docs/config_samples/ingester/ls5_pq_albers_s3.yaml
docs/config_samples/ingester/ls5_pq_albers_s3test.yaml
 - more sample ingest yamls for s3 and s3test.

omad committed:
Make s3 dependencies optional

omad committed:
More descriptive naming of driver methods

omad committed:
Make Multiprocessing Executor test a bit more robust to timing failures

omad committed:
Skip tests and pylint when S3 deps aren't available

omad committed:
Merge remote-tracking branch 'origin/develop' into csiro/s3-driver

omad commented:
Thanks @petewa , this is looking much closer to what we both had in mind.

I've gone further towards making the S3 dependencies optional by making the S3 specific tests `skip` if they can't be run.

Still outstanding is support for Windows filenames in the Driver Manager. When running on Windows there is still a failing few tests:
```
_______________ test_index_dataset_with_location[NetCDF CF-UTC] _______________

index = Index<db=PostgresDb<engine=Engine(postgresql://localhost:5432/agdcintegration)>>
default_metadata_type = MetadataType(name='eo', id_=1)
driver = <NetCDFDriver.NetCDFDriver object at 0x000000000F15DE48>

    def test_index_dataset_with_location(index, default_metadata_type, driver):
        """
        :type index: datacube.index._api.Index
        :type default_metadata_type: datacube.model.MetadataType
        """
        first_file = Path('/tmp/first/something.yaml').absolute()
        second_file = Path('/tmp/second/something.yaml').absolute()
        first_uri = driver.as_uri(first_file)
        second_uri = driver.as_uri(second_file)

        type_ = index.products.add_document(_pseudo_telemetry_dataset_type)
        dataset = Dataset(type_, _telemetry_dataset, uris=[first_uri], sources={})
        index.datasets.add(dataset)
        stored = index.datasets.get(dataset.id)
        assert stored.id == _telemetry_uuid
        # TODO: Dataset types?
        assert stored.type.id == type_.id
        assert stored.metadata_type.id == default_metadata_type.id
        if driver.uri_scheme == 'file':
>           assert stored.local_path == Path(first_file)
E           AssertionError: assert WindowsPath('.') == WindowsPath('C:/tmp/first/something.yaml')
E            +  where WindowsPath('.') = Dataset <id=4ec8fe97-e8b9-11e4-87ff-1040f381a756 type=ls8_telemetry location=.>.local_path
E            +  and   WindowsPath('C:/tmp/first/something.yaml') = Path(WindowsPath('C:/tmp/first/something.yaml'))
```

omad committed:
Add more future imports for py2

petewa committed:
Merge branch 'develop' into csiro/s3-driver

petewa committed:
datacube/drivers/driver.py
 - Windows path compability

petewa committed:
datacube/api/core.py
 - fixed check for pathos.threading.ThreadPool

omad commented:
Hi @petewa and everyone. I'm doing a last look over everything this afternoon, and am hoping to merge it in then. I want to work out why the test coverage has dropped, maybe some tests have been skipped.

Watch this space.

omad committed:
Undo accidental pytest skippage

omad committed:
Make tests skip instead of failing if deps are missing

omad committed:
Also skip doctests on ImportError
